reference/security.rst

```diff
@@ -0,0 +1,291 @@
+^^^^^^^^^^^^^^^^
+
+When performing login via the device flow, the Juju CLI will:
+
+1. Request from JAAS a URL and random code which the user can use to login.
+2. The CLI will then wait for a response from JAAS.
+3. JAAS polls the identity server, which in turn is waiting for the user to complete their login.
+
+A diagram depicting this login flow can be found 
+`here <https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow>`__.
+
+This process does not rely on browser redirects and so is not susceptible to traditional browser vulnerabilities.
+Both the access token and refresh token are obtained and stored by the back-end server
+and the CLI application is issued with a session token as described in `CLI Based Sessions`_.
+
+Client Credential Flow
+^^^^^^^^^^^^^^^^^^^^^^
+
+When there is a need to perform login by a machine rather than a physical user, OAuth handles this through
+the use of a client credential flow.
+
+A diagram depicting this login flow can be found 
+`here <https://auth0.com/docs/get-started/authentication-and-authorization-flow/client-credentials-flow>`__.
+
+This process is reserved for scenarios where machine to machine authentication is required. A good example
+of this is includes the use of the Juju Terraform Provider where the client-credential flow is employed.
+
+In the link above, the client application communicates with the identity provider to retrieve an access token
+that is then used with the API server. This design is shifted in JAAS. The client application instead sends
+its application credentials to JAAS which then forwards the information to the identity provider. JAAS
+effectively acts as a proxy between the client application and the identity provider.
+
+This scheme simplifies authentication for client applications but is only possible since JAAS is a trusted
+application in the system.
+
+Macaroons & Offer Authentication
+--------------------------------
+
+Macaroons are a tool for decentralised authentication similar to JSON Web Tokens.
+The `Go Macaroon package <https://pkg.go.dev/gopkg.in/macaroon.v2@v2.1.0>`__ is used by JAAS and has more
+details on the low-level operations that Macaroons are capable of.
+
+Macaroons are used by Juju for various purposes but in JAAS their primary purpose is for authorising 
+cross-model relations between controllers. When two Juju controllers (that are connected to JAAS) 
+communicate for the purposes of sharing an application offer, JAAS acts as the source of truth for 
+authorisation data. These checks are handled using macaroons.
+
+Macaroons use a combination of HMAC for cryptographic signatures and symmetric encryption to encode
+the scope (or caveats) of what a macaroon is entitled to.
+
+These operations are performed using ``HMAC-SHA256`` and ``XSalsa20-Poly1305``. The following Go 
+packages are used by the underlying macaroon package for these operations:
+
+- ``crypto/hmac``
+- ``crypto/sha256``
```

Comment:
**suggestion:** For a Go-unfamiliar reader, maybe it's better to remark that these two are standard library packages. What do you think?